File: src/share/classes/org/f3/runtime/Pointer.java
/*
 * Copyright 2008-2009 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

package org.f3.runtime;
import org.f3.runtime.sequence.Sequences;
import f3.animation.KeyValueTarget;
import org.f3.functions.*;

/**
 * Pointers
 *
 * @author Brian Goetz
 * @author A. Sundararajan
 */
public class Pointer<This extends F3Object, T> extends ConstPointer<This, T> implements KeyValueTarget<T>, f3.lang.MemberRef<This>, f3.lang.Ref<T>, f3.lang.ObservableRef<T> {
    @org.f3.runtime.annotation.F3Signature("(Ljava/lang/Object;)Lorg/f3/runtime/Pointer;")
    public static <This extends F3Object, T> Pointer<This,T> make(Type type, This obj, int varnum) {
        return new Pointer<This,T>(type, obj, varnum);
    }
    
    public static <This extends F3Object, T> Pointer<This,T> make(This obj, int varnum) {
        return new Pointer<This,T>(Type.OBJECT, obj, varnum);
    }
    
    public static boolean equals(Pointer p1, Pointer p2) {
        return (p1 == null) ? (p2 == null) : p1.equals(p2);
    }

    private Pointer(Type type, This obj, int varnum) {
	super(type, obj, varnum);
    }

    public void set(T value) {
        if (obj != null) {
            obj.set$(varnum, (Object)value);
        }
    }

    public Pointer unwrap() {
        return this;
    }
    
    public void set(int pos, T value) {
        assert type == Type.SEQUENCE : "expecting a sequence type";
        if (obj != null) {
	    Sequences.<T>set(obj, varnum, value, pos);
        }
    }

    public Object getValue() {
	return get();
    }

    public void setValue(Object value) {
	set((T)value);
    }

    public static void switchDependence(Pointer oldPtr, Pointer newPtr, F3Object dep, int depNum) {
        if (oldPtr != newPtr && dep != null) {
            F3Object oldSrc = (oldPtr != null)? oldPtr.getF3Object() : null;
            F3Object newSrc = (newPtr != null)? newPtr.getF3Object() : null;
            int oldVarNum = (oldPtr != null)? oldPtr.getVarNum() : 0;
            int newVarNum = (newPtr != null)? newPtr.getVarNum() : 0;
            dep.switchDependence$(oldSrc, oldVarNum, newSrc, newVarNum, depNum);
        }
    }

    /**
     * A BoundPointer is returned from the Pointer.bind(Pointer) method.
     * BoundPointer instance has to be kept alive till you want bind to be
     * effective. You can explicitly unbind the pointer using the "unbind"
     * method is this class.
     */
    public static class BoundPointer<a> extends Pointer<F3Object, a> {
        private Pointer srcPtr;
        private F3Object listener;

        private BoundPointer(Pointer<? extends F3Object,a> destPtr, Pointer<? extends F3Object,a> srcPtr, F3Object listener) {
            super(destPtr.getType(), destPtr.getF3Object(), destPtr.getVarNum());
            this.srcPtr = srcPtr;
            this.listener = listener;
        }

	public void set(a value) {
	    srcPtr.set(value);
	}

        /**
         * Uubind the current Pointer from the srcPtr. Repeated calls are fine
         * but subsequent calls are just no-ops. After unbind call, the BoundPointer
         * becomes effectively a regular, unbound Pointer. There is no need to
         * explicitly call 'unbind'. If this BoundPointer instance is unreachable,
         * GC will collect it and srcPtr will eventually see that the listener
         * object is not alive and so remove it from it's dependencies.
         */
        public void unbind() {
            srcPtr.removeDependency(listener);
            // clear everything related to Pointer bind.
            srcPtr = null;
            listener = null;
        }
    }

    /**
     * Implements identity bind expression between "srcPtr" and the current Pointer.
     * Whenever the value pointed by srcPtr changes, the current Pointer's value
     * is set from that.
     *
     * @param srcPtr The source Pointer object to which the current Pointer is bound to
     */
    public BoundPointer<T> bind(Pointer<? extends F3Object, T> srcPtr) {
        final F3Object thisObj = getF3Object();
        final int thisVarNum = getVarNum();
        final int srcVarNum = srcPtr.getVarNum();
        F3Object listener = new F3Base() {
            @Override
            public boolean update$(F3Object src, final int depNum,
				   int startPos, int endPos, int newLength, final int phase) {
                if ((phase & PHASE_TRANS$PHASE) == PHASE$TRIGGER) {
                    // update value from "src"
                    if (thisObj != null) {
                        thisObj.set$(thisVarNum, src.get$(srcVarNum));
                    }
                }
                return true;
            }
        };
        // initial update from "srcPtr"
        this.obj.set$(thisVarNum, (T)srcPtr.get());

        // add dependency so that we will get notified with update$ calls
        srcPtr.addDependency(listener);
        // return a BoundPointer so that use can call call "unbind()" later, if needed
        return new BoundPointer<T>(this, srcPtr, listener);
    }
}

File: f3_fmod/NativeFmodEx/Src/NativeFmodDesigner/Native/Pointer.cpp
/**
 * 							NativeFmodEx Project
 *
 * Do you want to use FMOD Ex API (www.fmod.org) with the Java language ? I've created NativeFmodEx for you.
 * Copyright © 2005-2007 Jérôme JOUVIE (Jouvieje)
 *
 * Created on 23 feb. 2005
 * @version file v1.0.0
 * @author Jérôme JOUVIE (Jouvieje)
 *
 *
 * WANT TO CONTACT ME ?
 * E-mail :
 * 		jerome.jouvie@gmail.com
 * My web sites :
 * 		http://jerome.jouvie.free.fr/
 * 		http://topresult.tomato.co.uk/~jerome/
 *
 *
 * INTRODUCTION
 * FMOD Ex is an API (Application Programming Interface) that allow you to use music
 * and creating sound effects with a lot of sort of musics.
 * FMOD is at :
 * 		http://www.fmod.org/
 * The reason of this project is that FMOD Ex can't be used direcly with Java, so I've created
 * this project to do this.
 *
 *
 * GNU LESSER GENERAL PUBLIC LICENSE
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of the License,
 * or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the
 * Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307 USA
 */

#include "Pointer.h"

jclass pointerClass = 0;
jclass getPointerClass(JNIEnv *jenv) {
	if(!pointerClass) {
		pointerClass = (jclass)jenv->NewGlobalRef(jenv->FindClass("org/jouvieje/FmodEx/Misc/Pointer"));
	}
	return pointerClass;
}
jmethodID newPointerID = 0;
jmethodID getNewPointerID(JNIEnv *jenv) {
	if(!newPointerID) {
		newPointerID = jenv->GetStaticMethodID(getPointerClass(jenv), "newPointer", "(J)Lorg/jouvieje/FmodEx/Misc/Pointer;");
	}
	return newPointerID;
}
jfieldID addressId = 0;
jfieldID getFieldID(JNIEnv *jenv) {
	if(!addressId) {
		addressId = jenv->GetFieldID(getPointerClass(jenv), "pointer", "J");
	}
	return addressId;
}

jobject newPointer(JNIEnv *jenv) {
	jobject result = jenv->CallStaticObjectMethod(getPointerClass(jenv), getNewPointerID(jenv), 0);
	return result;
}
jobject newPointer(JNIEnv *jenv, long address) {
	jobject result = jenv->CallStaticObjectMethod(getPointerClass(jenv), getNewPointerID(jenv), (jlong)address);
	return result;
}

long getPointerAddress(JNIEnv *jenv, jobject obj) {
	if(obj) {
		return (long)jenv->GetLongField(obj, getFieldID(jenv));
	}
	else {
		return 0;
	}
}

void setPointerAddress(JNIEnv *jenv, jobject obj, long newAddress) {
	if(obj) {
		jenv->SetLongField(obj, getFieldID(jenv), (jlong)newAddress);
	}
}

